synapse/events/snapshot.py

```diff
@@ -106,21 +106,9 @@ class EventContext(UnpersistedEventContextBase):
         state_delta_due_to_event: If `state_group` and `state_group_before_event` are not None
             then this is the delta of the state between the two groups.
 
-        prev_group: If it is known, ``state_group``'s prev_group. Note that this being
-            None does not necessarily mean that ``state_group`` does not have
-            a prev_group!
-
-            If the event is a state event, this is normally the same as
-            ``state_group_before_event``.
-
-            If ``state_group`` is None (ie, the event is an outlier), ``prev_group``
-            will always also be ``None``.
-
-            Note that this *not* (necessarily) the state group associated with
-            ``_prev_state_ids``.
-
-        delta_ids: If ``prev_group`` is not None, the state delta between ``prev_group``
-            and ``state_group``.
+        state_group_deltas: a dict containing information about the state groups associated with
+            the event and the state delta between these groups, ie a map from
+            (prev state group, new state group) -> delta state dict
```

Comment:
I'm not really sure what this dictionary contains, it seems to only ever get set with two entries in it, but I don't really get how those differ. Can an event have multiple previous state groups and multiple new ones?

Reply:
It's a little confusing because an event has two state groups: the one before the event and the one after the event. Each of those state groups will also have a previous state group. If the event is a state event, then the state group after the event will be different than the state group before the event - and the state group before the event will become the prev group for the state group after the event. 
If the event is not a state event, then the state group before the event and the state group after the event are the same, and the prev group for the state group before the event becomes the prev group for the state group after the event. 
We only ever care about the prev group for the state group after the event, and the hope is that by adding this field it is clear which prev group is associated with which state group and it's state delta (as they are contained in the dict together). Let know if this answers your question - I might have gone on a tangent here.

Reply:
I'm sorry, but I'm not following this. I'll try to reword and see if I'm anywhere close.

An event has a state group before and after the event. Each state group also has a previous state group, so we have two pairs of state groups:

* (State group before event, prev state group before event)
* (State group after event, prev state group after event)

For a state event the state group before the event becomes the new prev state group after event:

* (N, M)
* (P, N)

For a non-state event there is no change in the state groups:

* (N, M)
* (N, M)

Hopefully that's correct?

----

> We only ever care about the prev group for the state group after the event, and the hope is that by adding this field it is clear which prev group is associated with which state group and it's state delta

Then why even store the information from before the event? It seems we previously only store the information we need, which looks a lot clearer to me. I'm guessing this is a stepping stone to something else, but I'm finding the new code more confusing, what will this empower us to do in the future?

Reply:
Your understanding is correct, thanks for writing that out clearly. 

As for the second part, I think what I meant was not "we only care about the prev group for the state group after the event" what I really meant was "we only care about the prev group for the state group after the event and the state dict representing the state delta between state group before the event and the state group after the event", which is what is captured in the new field. 

Adding the `state_group_deltas` field was intended to relieve any future developers from having to parse out what the `prev_group` means, as this was also a complication when working with `UnpersistedEventContext`. I guess fundamentally the idea was to make the role of `prev_group` less confusing during the transition from `UnpersistedEventContext` --> `EventContext`. The `UnpersistedEventContext` does not yet have a state group after the event, so it only has the field `prev_group_for_state_group_before_event`. Rather than adding another field to the `UnpersistedEventContext` for `prev_group_for_state_group_after_the_event` the decision was made to add this all to the `state_group_deltas` instead, which represents the actual information we are after anyway (rather than building it from `prev_group` and `delta_ids` when persisting the event, see [here](https://github.com/matrix-org/synapse/blob/afb216c202feb143ce70c74c16fa50ca93da6157/synapse/storage/controllers/persist_events.py#L843)) and skip having to track all of this in the form of `prev_group`. 

I am sorry if this is hopelessly confusing, talking about `prev_group` was endlessly confusing for Erik and I and we thought this might be a way of saving others pain down the line. I am open if that's not the case, however!

Reply:
I'm just not finding this any clearer then the code is today -- it seems to be duplicating information we already have in the `_state_group`, `state_group_before_event`, and `state_delta_due_to_event` fields?

----

Looking at the changes again here, *most* of the changes are internal to `EventContext`, but they're then also in `persist_events`, in here we:

* Have a dictionary which always has only two items in it;
* Only ever pull a single item out in the persistence code path.

This has code smells to it in my mind. If we know we only need one of those why do we store both of them? If we know there's always two, why use a dictionary?

Reply:
Totally fair - let me go back to the drawing board with this and see if I can't come up with anything clearer.